internal/painter/software/draw.go

```diff
@@ -34,29 +32,33 @@ func drawImage(c fyne.Canvas, img *canvas.Image, pos fyne.Position, frame fyne.S
 	bounds := img.Size()
 	width := internal.ScaleInt(c, bounds.Width)
 	height := internal.ScaleInt(c, bounds.Height)
-	tex := painter.PaintImage(img, c, width, height)
 
 	if img.FillMode == canvas.ImageFillStretch {
 		width = internal.ScaleInt(c, img.Size().Width)
 		height = internal.ScaleInt(c, img.Size().Height)
-		tex = resize.Resize(uint(width), uint(height), tex, resize.Lanczos3)
 	} else if img.FillMode == canvas.ImageFillContain {
-		imgAspect := painter.GetAspect(img)
+		imgAspect := float32(img.Size().Width) / float32(img.Size().Height)
```

Comment:
Sorry - what you have calculated here is the output aspect ratio - what the `GetAspect` is returning is a cache of the source image ratio. With this change the `objAspect` below will always be 1, this is the source of all your test failures.